components/supervisor/frontend/src/ide/heart-beat.ts

```diff
@@ -70,6 +65,27 @@ export function schedule(wsInfo: WorkspaceInfo, sessionId: string): void {
         sendHeartBeat();
     }, activityInterval);
     toCancel.push(Disposable.create(() => clearInterval(intervalHandle)));
+
+    const handleUnload = async () => {
+        if (isSaaS) {
+            unloadListener = () => { sendHeartBeat(true) };
+            window.addEventListener('unload', unloadListener);
+            toCancel?.push(Disposable.create(() => window.removeEventListener('unload', unloadListener)));
```

Comment:
what if it already cancelled? Should we check for cancelled after each await? or add listener sync always and move condition within a listener?